### ISSUE TITLE
Rust integration

### ISSUE THREAD

**Issue by rlewkowicz** [roboflow/rf-detr#737]
### Search before asking

- [x] I have searched the RF-DETR [issues](https://github.com/roboflow/rf-detr/issues) and found no similar feature requests.

### Description

The headline is wall clock time is 2-4x better, with FAR less memory pressure, and more constant gpu utilization.  

I've done such an insane amount of optimization, it's hard to describe where I'm even trying to start. Part of this is me looking for feedback on exactly what parts anyone is interested in, if any. I'm hoping @isaacrob can give me critical suggestions. I know you're busy, if this is too much, don't feel obligated to read all this, `---` ends critical question.

The biggest change is that all the dataloading is now done in a rust module, and it packs tensors on the gpu, not just prefetching in the cpu worker, but also pre batching in the gpu. All images are precompiled to disk at inference size. This is not supposed to be a general purpose loader. It's an opinionated, defined loader that hands off via zero copy into python. Stays loaded between val/train phases, should be DDP compliant without memory scaling. At this point, the python framework is just a book keeper running the model and python dataloaders are disabled. 

You don't want to build a rust module in this repo I imagine? What about a plugin interface of some sort? Would you allow a custom plugin and launch argument? I will nail down the specifics to ensure a professional well articulated pull but if I can create a clean, easy to install dependency, can I at least get an optional in the pyproject, and a blurb in the README.md? 

---
Then here's just some additional tweaks I made, each number would probably merit it's own pull request. That data loader should probably be light touch if I get the interfaces right. These are all explicit changes to the code base. We can gate them, but they add up
1. Dataloader changes sets the stage for basic dynamo compilation given a fast loading, precompiled, fixed shape dataset.  I had to go add a handful of decorators for functions that had changing tensor shapes so they compile etc. Its broken graph allowed, but that helps a lot. Little flags for "fast math" and other cuda improvements. 
2. Broad time complexity improvements. There was smaller testing and benchmarks, but not everything was atomically tested, just broader outcomes. Included among them threading that Hungarian matcher and some other loss function optimizations. 
3. General threading of blocking tasks. 

I'd have to go clean all this stuff up, get it production ready, clean, incremental etc etc. Do we even want any of it? 

Once this is compiled, it's insanely fast. I've tried a lot, I have funky memory issues with pytorch and large datasets across python and any application. This is fuzzy math, but I believe I was at ~2:20 an epoch, I split my dataset into 3rds, so its about the same as the coco in training length. I've added 25% more images, I add a whole preaug set. So now with a training set 25% larger I'm at   ~1:04. Thats just flat train time. I've pinned workers, but there's a huge delay between phases etc. It's instant comparatively. There is next to no downtime between phases. 

### Use case

Overall optimization 

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by isaacrob** [comment]
@Matvezy is also cleaning up the Hungarian matcher, getting meaningful speedup :) should be fairly optimized soon. lots of places in this code base where things can be improved!

the big changes are doing better threading and improving the data loader? my main question is, in what cases would one prefer this dataloader vs torch's built-in one, which is already supposed to be quite optimized? would you consider making a standalone repo that allows people to use it as a drop-in replacement for torch's version?

@Borda is leading a major refactor of this repo to be built on top of pytorch lightning, see https://github.com/roboflow/rf-detr/issues/709. I am expecting significant performance improvements to come as a result of that. would you expect your changes to still be relevant in that case?

one big thing I'm after is being able to torch compile end to end the whole training loop. you made reference to something like that, can you expand?

I would defer to @Borda and @SkalskiP on the tradeoffs of having a rust dependency

**Comment by rlewkowicz** [comment]
@isaacrob @Matvezy @Borda @SkalskiP 

Sorry about the delay. I just kept going in at it, trying to not modify the source too much. It got gross, and at the end of the day, I need the speed. 

https://github.com/rlewkowicz/cpp-rf-detr

This is a private repo (I added yall you might have to accept in the github console), idk what I'm doing with this yet. This is for yall to take and use as you please. I gotta add a license (just for roboflow folks though). Technically, this is my competitive advantage.  

I don't think you'll ever come close with python. I played with lightning. It was ok. This is a full training platform. 

~105k images in 1h 20m, averaging 64 annotations per image. So it's dense masks too. I think your py defaults are double val, and val loss, so this is a little disingenuous, but validation is like 5-7 min (should be 2-3 min, I'm fixing that), something like 10x less my python runs.

<img width="2056" height="816" alt="Image" src="https://github.com/user-attachments/assets/c181bca5-abeb-46fe-a6dc-dbf039dd1690" />

I introduce a concept of lanes which is very similar to a trt context with multiple execution streams. I'm going to try to validate DDP today. This will never be multinode, but my lanes concept should allow me to scale gpu util to bigger systems. GPU on my 3090ti stays pegged at ~90% and that epoch time and b/s stays smoooooooth. 

You all know how brutally time consuming this is and I aint got the compute to be playin. So the smoke test is a saturated seg med model, 95% box, 87% seg. Continued for 5-10 epochs with expected map/loss fluctuations. No degradation (and even some mild gains).

The only kicker is the dataset. My dataset is superior for a number of reasons (should I choose to try to build a online product, there's reasons it's a better architecture). Pytorch compile, 20min. My compile? 2-4 min. Start time? IMMEDIATE. Run the command, no loading, no waiting, it just goes. 

This should build nice and clean on a fedora 43 system. I'll finish docker, DDP, and I'll create a coco set in my format. But on the "trust me bro" metric, this should be a mathematically equivalent training platform. 

At the end of the day, if you are an organization selling compute, and I say I can do this with 2-10x less compute, there's no competition. 1:20 on seg med vs (again a little disingenuous I think you have rough defaults), what? 4-5 hours? Not including hangs between epochs. Not including val. So 6-7 hours per epoch? I'm so much faster it's crazy. Memory pressure? Forget about it. I'm running at 20 gigs with a comparable data set running 60 to 100 gigs in python. Mmm. Delicious.

If yall like this, I would help with something (or take it and use it as you please). I want to build a whole distributed system with this and HBASE. If I can pump out models in a 10th of the time, thats money. 

(On Hungarian. Your big win is just not blocking per image, parallelize the batch if I recall, that's a simple win)
